### PR TITLE
feat(jotai): add @suspensive/jotai package

### PR DESCRIPTION
**Pull request by gwansikk**
# Overview

## Roadmap

- [x] Atom (setAtom)
- [x] SetAtom (useSetAtom)
- [x] AtomValue (useAtomValue)

## Tests

- [x] local package test (pack .tgz)
- [x] jotai-tanstack-query (query, mutation)

Jul-07-2024 21-34-16

## PR Checklist

- [x] I did below actions if need

1. I read the [Contributing Guide](https://github.com/toss/suspensive/blob/main/CONTRIBUTING.md)
2. I added documents and tests.

### CODE REVIEW

File: packages/jotai/package.json [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  "homepage": "https://suspensive.org",
+  "bugs": "https://github.com/toss/suspensive/issues",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/toss/suspensive.git",
+    "directory": "packages/jotai"
+  },
+  "license": "MIT",
+  "author": {
+    "name": "GwanSik Kim and Jonghyeon Ko",
```
Comment (manudeli) [inline packages/jotai/package.json:18]:
```suggestion
    "name": "GwanSik Kim & Jonghyeon Ko",
```

File: packages/jotai/src/AtomValue.test-d.tsx [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+describe('<AtomValue/>', () => {
+  it('type check', () => {
+    ;() => (
+      <AtomValue atom={asyncAtom}>
+        {([value]) => {
+          expectTypeOf(value).toEqualTypeOf<string>()
+          return <></>
+        }}
+      </AtomValue>
+    )
```
Comment (manudeli) [inline packages/jotai/src/AtomValue.test-d.tsx:16]: https://jotai.org/docs/core/use-atom#useatomvalue

In my opinion, This should be just value, not array destructuring

```suggestion
describe('<AtomValue/>', () => {
  it('type check', () => {
    ;() => (
      <AtomValue atom={asyncAtom}>
        {(value) => {
          expectTypeOf(value).toEqualTypeOf<string>()
          return <></>
        }}
      </AtomValue>
    )
```

**Review by manudeli — CHANGES_REQUESTED** [review]

File: packages/jotai/src/types.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import type { WritableAtom } from 'jotai'
+import type { ReactNode } from 'react'
+
+export interface WithChildren<TValue> {
+  children: (value: TValue) => ReactNode
+}
```
Comment (manudeli) [inline packages/jotai/src/types.ts:6]: Name could be more intuitive.

```suggestion
export interface CallbackChildrenProps<TValue> {
  children: (value: TValue) => ReactNode
}
```

File: packages/jotai/src/types.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+export type AtomProps<TAtom, TOptions = unknown> = {
+  atom: TAtom
+  options?: TOptions
+}
```
Comment (manudeli) [inline packages/jotai/src/types.ts:11]: AtomProps should be ComponentProps for Atom component

```suggestion
```

**Review by manudeli — CHANGES_REQUESTED** [review]

**Comment by manudeli** [comment]
@gwansikk I will call you after making new Pull Request into this branch to modify this codes. could you review it later?

File: packages/jotai/src/AtomValue.test-d.tsx (hunk 2 above)
Comment (gwansikk) [reply packages/jotai/src/AtomValue.test-d.tsx:16]: That’s correct. My mistake. Thank you!

File: packages/jotai/package.json [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  ],
+  "homepage": "https://suspensive.org",
+  "bugs": "https://github.com/toss/suspensive/issues",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/toss/suspensive.git",
+    "directory": "packages/jotai"
+  },
+  "license": "MIT",
+  "author": "GwanSik Kim <Seorit98@gmail.com> & Jonghyeon Ko <jonghyeon@toss.im>",
```
Comment (gwansikk) [inline packages/jotai/package.json:17]: 🆒👍

File: packages/jotai/src/Atom.tsx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  >): ReactNode
+
+/**
+ * @experimental This is experimental feature.
+ */
+export function Atom<TAtom extends JotaiAtom<unknown>>({
+  children,
+  atom,
+  options,
+}: AtomProps<TAtom> & ChildrenRenderProps<AtomResult<Awaited<ExtractAtomValue<TAtom>>, never>>): ReactNode
```
Comment (gwansikk) [inline packages/jotai/src/Atom.tsx:73]: - https://github.com/pmndrs/jotai/blob/2c021f9092abb47611ff0c471738a666d011ae55/src/react/useAtom.ts#L45
- https://jotai.org/docs/guides/async#async-read-atom

Changed the `TSetter` type to be inferred as `never` for async read atoms. (Previously it was `unknown`)

File: packages/jotai/src/Atom.tsx [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+/**
+ * @experimental This is experimental feature.
+ */
+export function Atom<TValue, TArgs extends unknown[], TResult>({
+  children,
+  atom,
+  options,
+}: AtomProps<JotaiAtom<unknown>> & ChildrenRenderProps<any>): ReactNode {
```
Comment (gwansikk) [inline packages/jotai/src/Atom.tsx:82]: To include the `never` type and support function overload, type widening (`any`) has been applied.

**Comment by manudeli** [comment]
https://github.com/toss/suspensive/pull/1045 I left improving Pull Request

**Review by manudeli — APPROVED** [review]
@gwansikk Could I merge first? In my opinion, interfaces should be improved and unified. But in zero version @suspensive/jotai could be improved even with BREAKING CHANGE. so It would be better that we release it first

**Comment by gwansikk** [comment]
> @gwansikk Could I merge first? In my opinion, interfaces should be improved and unified. But in zero version @suspensive/jotai could be improved even with BREAKING CHANGE. so It would be better that we release it first
> 
> 
> 
> In my opinion, It is better to merge first to create an environment where documentation and feature improvements can be worked on in parallel.

Yes, I agree! It seems better to deploy version 0.0.0 first and then make improvements.

**Comment by manudeli** [comment]
@gwansikk I added @suspensive/jotai in codecov.yml at [b10b946](https://github.com/toss/suspensive/pull/1042/commits/b10b9467a968f277956086bb1a15998302eaab8e)

**Pull request merged**